server/autolink/autolink.go

```diff
@@ -108,6 +108,12 @@ func (l Autolink) Replace(message string) string {
 		if submatch == nil {
 			break
 		}
+
+		// The beginning of the submatch is equal to the end of the submatch here. The regex pattern is non-terminal.
```

Comment:
Why is this true? Can you explain what is happening here? I don't quite understand.

Reply:
@crspeller @levb

Before the loop begins, the post's message is converted to a `[]byte` and stored in the local variable `in`. Another `[]byte` is created to store the output of the replace process.

In the loop, the `regexp` package's [FindSubmatchIndex](https://pkg.go.dev/regexp#Regexp.FindSubmatchIndex) function is called with the input array, and the first submatch found is processed. The `FindSubmatchIndex` function returns a `[]int` that contains indexes of submatches. Each pair of two entries in the array represents the start and end index of each submatch (exclusive, meaning the end index will be equal to the length of the match if the start index is 0). The exit condition of the loop is when `FindSubmatchIndex` returns nil, which only happens when no match is found.

For the autolink's algorithm, only the first submatch matters at any given time, because the algorithm processes the submatch, **and then moves its cursor forward in the `in` array and sees if there are more submatches to process in the next iteration of the loop**. Because of this, only the first two elements in `FindSubmatchIndex`'s return value are ever important to the autolink replace algorithm at any given time.


When the autolink algorithm is provided a regex pattern such as `.*`:

The first run through the loop causes the entire input string to match. The first element in `FindSubmatchIndex`'s return value is `0` and second is the length of the message. Line 119 results in the `in` array being empty now, since the whole string was matched and `submatch[1] == len(in)` in this case. Note that there is no condition checked at the end of the loop to see if we should leave the loop for any reason. The only condition to leave is "did `FindSubmatchIndex` return nil?" at the beginning of the loop.

When we run through the loop again, the argument given to `FindSubmatchIndex` is an empty array, and it returns the value `[0, 0]`, meaning "the match starts at index zero, and ends at index zero", which means it matched on an empty string. Since `submatch` is not nil here, the loop moves on, though results in a no-op for everything else below in the loop since `in` is empty. The loop continues and does the same calculation over and over, finding the same empty submatch from the `.*` pattern indefinitely.

The solution I've made here is "if a submatch is found but has no length, we have nothing to process, so we should exit the loop." A better solution may be to check if `len(in) == 0` at the end of the loop.

Reply:
Thanks for the explanation! I think you are right that checking the length would be better.  Maybe more intuitive at the beginning of the loop?